lib/compat/wordpress-6.0/class-wp-webfonts.php

```diff
@@ -62,10 +79,264 @@ public function init() {
+
+		if ( ! $webfonts_used_in_global_styles ) {
+			$webfonts_used_in_global_styles = $this->collect_webfonts_used_in_global_styles();
+
+			if ( ! defined( 'WP_DEBUG' ) || ! WP_DEBUG ) {
+				update_post_meta( $global_styles_post_id, self::$webfonts_cache_meta_attribute, $webfonts_used_in_global_styles );
+			}
+		}
+
+		$webfont_ids = array_keys( $webfonts_used_in_global_styles );
+
+		foreach ( $webfont_ids as $webfont ) {
+			if ( isset( $this->get_registered_fonts()[ $webfont ] ) ) {
+				$this->enqueue_font( $webfont );
+			}
+		}
+	}
+
+	/**
+	 * Collect globally used webfonts.
+	 *
+	 * @return array
+	 */
+	private function collect_webfonts_used_in_global_styles() {
+		$global_styles                  = gutenberg_get_global_styles();
+		$webfonts_used_in_global_styles = array();
+
+		if ( isset( $global_styles['blocks'] ) ) {
+			// Register used fonts from blocks.
+			foreach ( $global_styles['blocks'] as $setting ) {
+				$font_family_slug = $this->get_font_family_from_setting( $setting );
+
+				if ( $font_family_slug ) {
+					$webfonts_used_in_global_styles[ $font_family_slug ] = 1;
+				}
+			}
+		}
+
+		if ( isset( $global_styles['elements'] ) ) {
+			// Register used fonts from elements.
+			foreach ( $global_styles['elements'] as $setting ) {
+				$font_family_slug = $this->get_font_family_from_setting( $setting );
+
+				if ( $font_family_slug ) {
+					$webfonts_used_in_global_styles[ $font_family_slug ] = 1;
+				}
+			}
+		}
+
+		// Get global font.
+		$font_family_slug = $this->get_font_family_from_setting( $global_styles );
+
+		if ( $font_family_slug ) {
+			$webfonts_used_in_global_styles[ $font_family_slug ] = 1;
+		}
+
+		return $webfonts_used_in_global_styles;
+	}
+
+	/**
+	 * Get font family from global setting.
+	 *
+	 * @param mixed $setting The global setting.
+	 * @return string|false
+	 */
+	private function get_font_family_from_setting( $setting ) {
+		if ( isset( $setting['typography'] ) && isset( $setting['typography']['fontFamily'] ) ) {
+			$font_family = $setting['typography']['fontFamily'];
+
+			preg_match( '/var\(--wp--(?:preset|custom)--font-family--([^\\\]+)\)/', $font_family, $matches );
+
+			if ( isset( $matches[1] ) ) {
+				return _wp_to_kebab_case( $matches[1] );
+			}
+
+			preg_match( '/var:(?:preset|custom)\|font-family\|(.+)/', $font_family, $matches );
+
+			if ( isset( $matches[1] ) ) {
+				return _wp_to_kebab_case( $matches[1] );
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Invalidate template webfonts cache.
+	 */
+	public function invalidate_template_webfonts_cache() {
+		remove_theme_mod( self::$template_webfonts_cache_theme_mod );
+	}
+
+	/**
+	 * Hook into every possible template so we can enqueue the webfonts used in the template
+	 * that has been loaded in the front-end.
+	 */
+	public function register_filter_for_current_template_webfonts_enqueuing() {
+		$template_type_slugs = array_keys( get_default_block_template_types() );
```

Comment:
Will this work for custom templates?

Reply:
I think it will only work for custom templates with the same slugs as those defaults, but I think would miss others that could be created by users or developers (like header-footer-only for example isnt part of the default list.

Are we able to loop through all `wp_templates` without relying on a list of possible slugs?

Reply:
I wonder if `gutenberg_get_block_templates` would work to create a slug list like this from? I think that should be retrieving all templates both from the theme file structure and the database.

Reply:
So, if I got this right, `get_default_block_template_types` returns a list of the "slots" where the templates are applied. If this is right, then I think we should be good to go -- we want to hook into the rendered template and scan them for fonts as lazily as possible.

Reply:
`gutenberg_get_block_templates` retrieves all templates, whether they're rendered or not, and that might be costly because existing doesn't mean that the template is actually being used. I think a good idea would be to get as lazy as it gets.